hbase-procedure/src/main/java/org/apache/hadoop/hbase/procedure2/ProcedureSuspendedException.java

```diff
@@ -35,5 +38,21 @@ public ProcedureSuspendedException() {
    */
   public ProcedureSuspendedException(String s) {
     super(s);
+    shouldForceLockRelease = false;
+  }
+
+  /**
+   * Constructor
+   * @param s                      message
+   * @param shouldForceLockRelease if true, the procedure will release the lock regardless of
+   *                               {@link Procedure#holdLock(Object)}
+   */
+  public ProcedureSuspendedException(String s, boolean shouldForceLockRelease) {
+    super(s);
+    this.shouldForceLockRelease = shouldForceLockRelease;
+  }
+
+  public boolean shouldForceLockRelease() {
```

Comment:
The Procedure class has a `holdLock` method which returns true for SnapshotProcedure. This prevents us from releasing the lock even on suspension. I believe that at the stage where the deadlock could happen, it's safe to release the lock. We haven't done any snapshotting yet, so I believe there shouldn't be any weird data inconsistencies from region splits/merges.

Reply:
Shouldn't we just modify the SnapshotProcedure.holdLock() behaviour? Afterall, this is contrary to what we say in that method comment:

`  protected boolean holdLock(MasterProcedureEnv env) {
    // In order to avoid enabling/disabling/modifying/deleting table during snapshot,
    // we don't release lock during suspend
    return true;
  }
`
And you are already modifying SnapshotProcedure execution logic to reflect the special condition when the lock should be released, so better keep this there and avoid changing ProcedureSuspendedException?